tsdb/aggregate.go

```diff
@@ -0,0 +1,892 @@
+		tMins := make(int64Slice, 0, len(buckets))
+		for k, _ := range buckets {
+			tMins = append(tMins, k)
+		}
+
+		if ascending {
+			sort.Sort(tMins)
+		} else {
+			sort.Sort(sort.Reverse(tMins))
+		}
+
+		values := make([][]interface{}, len(tMins))
+		for i, t := range tMins {
+			values[i] = make([]interface{}, 0, len(columnNames))
+			values[i] = append(values[i], time.Unix(0, t).UTC()) // Time value is always first.
+
+			for j, f := range reduceFuncs {
+				reducedVal := f(buckets[t][j])
+				values[i] = append(values[i], reducedVal)
+			}
+		}
+
+		// Perform aggregate unwraps
+		values, err = e.processFunctions(values, columnNames)
+		if err != nil {
+			out <- &models.Row{Err: err}
+		}
+
+		// Perform any mathematics.
+		values = processForMath(e.stmt.Fields, values)
+
+		// Handle any fill options
+		values = e.processFill(values)
+
+		// process derivatives
+		values = e.processDerivative(values)
+
+		// If we have multiple tag sets we'll want to filter out the empty ones
+		if hasMultipleTagSets && resultsEmpty(values) {
+			continue
+		}
+
+		row.Values = values
+		out <- row
+	}
+
+	close(out)
+}
+
+// initReduceFuncs returns a list of reduce functions for the aggregates in the query.
+func (e *AggregateExecutor) initReduceFuncs() ([]reduceFunc, error) {
+	calls := e.stmt.FunctionCalls()
+	fns := make([]reduceFunc, len(calls))
+	for i, c := range calls {
+		fn, err := initializeReduceFunc(c)
+		if err != nil {
+			return nil, err
+		}
+		fns[i] = fn
+	}
+	return fns, nil
+}
+
+// openMappers opens all the mappers.
+func (e *AggregateExecutor) openMappers() error {
+	for _, m := range e.mappers {
+		if err := m.Open(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// initMappers buffers the first chunk of each mapper.
+func (e *AggregateExecutor) initMappers() error {
+	for _, m := range e.mappers {
+		chunk, err := m.NextChunk()
+		if err != nil {
+			return err
+		}
+		m.bufferedChunk = chunk
+
+		if m.bufferedChunk == nil {
+			m.drained = true
+		}
+	}
+	return nil
+}
+
+// hasMultipleTagSets returns true if there is more than one tagset in the mappers.
+func (e *AggregateExecutor) hasMultipleTagSets() bool {
+	set := make(map[string]struct{})
+	for _, m := range e.mappers {
+		for _, t := range m.TagSets() {
+			set[t] = struct{}{}
+			if len(set) > 1 {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// ascending returns true if statement is sorted in ascending order.
+func (e *AggregateExecutor) ascending() bool {
+	if len(e.stmt.SortFields) == 0 {
+		return true
+	}
+	return e.stmt.SortFields[0].Ascending
+}
+
+// mappersDrained returns whether all the executors Mappers have been drained of data.
+func (e *AggregateExecutor) mappersDrained() bool {
+	for _, m := range e.mappers {
+		if !m.drained {
+			return false
+		}
+	}
+	return true
+}
+
+// nextMapperTagset returns the alphabetically lowest tagset across all Mappers.
+func (e *AggregateExecutor) nextMapperTagSet() string {
```

Comment:
I presume not, just to decide on the order of the tagsets to spit out. Just wanted to be sure.